File: VialerTests/RecentCallTests.m
//
//  RecentCallTests.m
//  Copyright © 2016 VoIPGRID. All rights reserved.
//

#import <CoreData/CoreData.h>
#import <OCMock/OCMock.h>
#import "RecentCall.h"
@import XCTest;

@interface RecentCallTests : XCTestCase
@property (strong, nonatomic) NSManagedObjectContext *moc;
@end

@implementation RecentCallTests

- (void)setUp {
    [super setUp];
    NSURL *modelURL = [[NSBundle mainBundle] URLForResource:@"VialerModel" withExtension:@"momd"];
    NSManagedObjectModel *mom = [[NSManagedObjectModel alloc] initWithContentsOfURL:modelURL];
    NSPersistentStoreCoordinator *psc = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:mom];
    XCTAssertTrue([psc addPersistentStoreWithType:NSInMemoryStoreType configuration:nil URL:nil options:nil error:NULL] ? YES : NO, @"Should be able to add in-memory store");
    self.moc = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
    self.moc.persistentStoreCoordinator = psc;
}

- (void)tearDown {
    self.moc = nil;
    [super tearDown];
}

- (void)testGetLatestCallWillReturnTheCall {
    // Given
    RecentCall *recent = [NSEntityDescription insertNewObjectForEntityForName:@"RecentCall" inManagedObjectContext:self.moc];

    // When
    RecentCall *fetchedRecent = [RecentCall latestCallInManagedObjectContext:self.moc];

    //Then
    XCTAssertEqualObjects(recent, fetchedRecent, @"The correct RecentCall should have been returned");
}

- (void)testLatestCallReturnsNilWhenDatabaseEmpty {
    RecentCall *fetchedRecent = [RecentCall latestCallInManagedObjectContext:self.moc];
    XCTAssertNil(fetchedRecent, @"Database should have been empty, no object should have been returned");
}

- (void)testRecentInboundCallIsSuppressed {
    // Given
    RecentCall *recent = [NSEntityDescription insertNewObjectForEntityForName:@"RecentCall" inManagedObjectContext:self.moc];

    // When
    recent.inbound = @YES;
    recent.sourceNumber = @"+31222xxxxxx";

    // Then
    XCTAssertTrue([recent suppressed], @"The given inbound recent should have indicated that it is suppressed");
}
- (void)testRecentInboundCallButNotASuppressedNumber {
    //Given
    RecentCall *recent = [NSEntityDescription insertNewObjectForEntityForName:@"RecentCall" inManagedObjectContext:self.moc];

    // When
    recent.inbound = @YES;
    recent.sourceNumber = @"+31222333333";

    // Then
    XCTAssertFalse([recent suppressed], @"The given inbound recent should NOT indicated as being suppressed");
}

- (void)testRecentOuboundCallCannotIndicateBeingSuppressed {
    // Given
    RecentCall *recent = [NSEntityDescription insertNewObjectForEntityForName:@"RecentCall" inManagedObjectContext:self.moc];

    // When
    recent.inbound = @NO;
    recent.sourceNumber = @"+31222xxxxxx";

    // Then
    XCTAssertFalse([recent suppressed], @"An outbound call can never indicate as being suppressed.");
}

- (void)testRecentInboundCallIsNotSuppressed {
    // Given
    RecentCall *recent = [NSEntityDescription insertNewObjectForEntityForName:@"RecentCall" inManagedObjectContext:self.moc];

    // When
    recent.inbound = @NO;
    recent.sourceNumber = @"+31222333333";

    // Then
    XCTAssertFalse([recent suppressed], @"The given recents should not indicate that it is suppressed");
}

@end

File: VialerTests/Recents/Models/RecentCallTests.swift
//
//  RecentCallTests.swift
//  Copyright © 2017 VoIPGRID. All rights reserved.
//

import XCTest
@testable import Vialer

class RecentCallTests: XCTestCase {

    let exampleDictionary: JSONDictionary = ["src_number": "+3150809000",
                                            "callerid": "VoIPGRID",
                                            "call_date": "2017-03-15T09:41:27",
                                            "direction": "inbound",
                                            "dst_number": "+31508009064",
                                            "dialed_number": "+31508009064",
                                            "atime": Int16(42),
                                            "id": Int64(123456789)]
    var coreDataStack: CoreDataStack!
    override func setUp() {
        super.setUp()
        coreDataStack = CoreDataStack(modelNamed: "Vialer", inMemory: true)
    }
    
    override func tearDown() {
        super.tearDown()
        coreDataStack = nil
    }

    func testRecentInboundCallIsSuppressed() {
        var dictionary = exampleDictionary
        dictionary["src_number"] = "xxxxxxx"
        let call = RecentCall.findOrCreate(for: dictionary, in: coreDataStack.mainContext)!

        XCTAssertTrue(call.suppressed, "Call should be suppressed")
    }

    func testRecentInboundCallIsNotSuppressed() {
        let call = RecentCall.findOrCreate(for: exampleDictionary, in: coreDataStack.mainContext)!

        XCTAssertFalse(call.suppressed, "Call should not be suppressed")
    }

    func testOutboundCallIsNeverSuppressed() {
        var dictionary = exampleDictionary
        dictionary["src_number"] = "xxxxxxx"
        dictionary["direction"] = "outbound"
        let call = RecentCall.findOrCreate(for: dictionary, in: coreDataStack.mainContext)!

        XCTAssertFalse(call.suppressed, "Call should not be suppressed")
    }

    func testOutboundCallIsNotSuppressed() {
        var dictionary = exampleDictionary
        dictionary["direction"] = "outbound"
        let call = RecentCall.findOrCreate(for: dictionary, in: coreDataStack.mainContext)!

        XCTAssertFalse(call.suppressed, "Call should not be suppressed")
    }

    func testNoDuplicateCallsWillBeCreated() {
        let call = RecentCall.findOrCreate(for: exampleDictionary, in: coreDataStack.mainContext)!
        let secondCall = RecentCall.findOrCreate(for: exampleDictionary, in: coreDataStack.mainContext)!

        XCTAssertEqual(call, secondCall, "Only one RecentCall should have been created.")
    }
}
